zcash_test_vectors/transaction_zsa.py

```diff
@@ -0,0 +1,157 @@
+import struct
+
+from .orchard.key_components import FullViewingKey, SpendingKey
+from .orchard_zsa.key_components import IssuanceKeys
+from .orchard.pallas import Fp as PallasBase, Point
+from .orchard.sinsemilla import group_hash as pallas_group_hash
+from .orchard_zsa.asset_base import zsa_value_base, asset_digest, encode_asset_id, get_random_unicode_bytes
+from .utils import leos2ip
+from .zc_utils import write_compact_size
+from .transaction import (
+    MAX_MONEY, NOTEENCRYPTION_AUTH_BYTES,
+    ZC_SAPLING_ENCPLAINTEXT_SIZE, ZC_SAPLING_OUTCIPHERTEXT_SIZE,
+    OrchardActionBase,
+    RedPallasSignature,
+    TransactionBase,
+)
+
+NU7_VERSION_GROUP_ID = 0x124A69F8
+NU7_TX_VERSION = 7
+
+# Orchard ZSA note values
+ZC_ORCHARD_ZSA_ASSET_SIZE = 32
+ZC_ORCHARD_ZSA_ENCPLAINTEXT_SIZE = ZC_SAPLING_ENCPLAINTEXT_SIZE + ZC_ORCHARD_ZSA_ASSET_SIZE
+ZC_ORCHARD_ZSA_ENCCIPHERTEXT_SIZE = ZC_ORCHARD_ZSA_ENCPLAINTEXT_SIZE + NOTEENCRYPTION_AUTH_BYTES
+
+class OrchardZSAActionDescription(OrchardActionBase):
+    def __init__(self, rand):
+        super().__init__(ZC_ORCHARD_ZSA_ENCCIPHERTEXT_SIZE, rand)
+
+class AssetBurnDescription(object):
+    def __init__(self, rand):
+        isk = IssuanceKeys(rand.b(32))
+        desc_size = rand.u32() % 512 + 1
+        desc_bytes = get_random_unicode_bytes(desc_size, rand)
+        asset_digest_bytes = asset_digest(encode_asset_id(isk.ik, desc_bytes))
+        self.assetBase : Point = zsa_value_base(asset_digest_bytes)
+        self.valueBurn = rand.u64()
+
+    def __bytes__(self):
+        return bytes(self.assetBase) + struct.pack('<Q', self.valueBurn)
+
+class IssueActionDescription(object):
+    def __init__(self, rand, ik):
+        self.assetDescSize = rand.u32() % 512 + 1
+        self.asset_desc = get_random_unicode_bytes(self.assetDescSize, rand)
+        self.vNotes = []
+        for _ in range(rand.u8() % 5):
+            self.vNotes.append(IssueNote(rand, ik, self.asset_desc))
+        self.flagsIssuance = rand.u8() & 1    # Only one bit is reserved for the finalize flag currently
+
+    def __bytes__(self):
+        ret = b''
+
+        ret += write_compact_size(self.assetDescSize)
+        ret += bytes(self.asset_desc)
+        ret += write_compact_size(len(self.vNotes))
+        if len(self.vNotes) > 0:
+            for note in self.vNotes:
+                ret += bytes(note)
+        ret += struct.pack('B', self.flagsIssuance)
+
+        return ret
+
+class IssueNote(object):
+    def __init__(self, rand, ik, asset_desc):
+        fvk_r = FullViewingKey.from_spending_key(SpendingKey(rand.b(32)))
+        self.recipient = fvk_r.default_d() + bytes(fvk_r.default_pkd())
```

Comment:
probably both components need `bytes()` or none at all. No?

Reply:
let's keep the way it is

Reply:
Okay, done.